### PR TITLE
Chinese translation  domain driven design implementation guide

### PR DESCRIPTION
**Pull request by akinix**

### CODE REVIEW

**Comment by realLiangshiwei** [comment]
Can you add navigation to the `docs-nav.json`?

**Comment by rqx110** [comment]
有些文字笔误需要修正

File: docs/zh-Hans/Domain-Driven-Design-Implementation-Guide.md [hunk 1]
```diff
@@ -0,0 +1,1979 @@
...
+
+但是我们仍然建议采用仓储模式来隐藏基础设施中实现过程。
+
+#### 展现层技术无关原则
+
+展现层技术（UI框架）时现代应用程序中最多变的部分之一。**领域层和应用层**应该对展现层所采用的技术或框架**一无所知**。使用ABP启动模板就非常容易实现此原则。
+
+在某些情况下，您可能需要在应用层和展现层中写重复的逻辑，例如，参数验证和授权检查。展现层检查出于**用户体验**，应用层或领域层检查出于**数据安全性**和**数据完整性**。
+
+#### 关注状态的变化而报表/查询(个人觉得更适合)
```
Comment (rqx110) [inline docs/zh-Hans/Domain-Driven-Design-Implementation-Guide.md]: 原文`Focus on the State Changes, Not Reporting`愿意应该是`关注状态变化, 而不是XXX`. 有没有更好的翻译？

File: docs/zh-Hans/Domain-Driven-Design-Implementation-Guide.md [hunk 2]
```diff
@@ -0,0 +1,1979 @@
...
+}
+````
+
+### 数据传输对象
+
+[DTO](Data-Transfer-Objects.md)是应用层与展现层间传输数据的简单对象。应用服务方法获取并返回Dto。
+
+#### DTO通用原则和最佳实践
+
+* DTO应该是**可被序列化**的。因为大所属情况下，DTO是通过网络传输的，因此它应该具有**无参的构造函数**。
```
Comment (rqx110) [inline docs/zh-Hans/Domain-Driven-Design-Implementation-Guide.md]:
```suggestion
* DTO应该是**可被序列化**的。因为大多数情况下，DTO是通过网络传输的，因此它应该具有**无参的构造函数**。
```

File: docs/zh-Hans/Domain-Driven-Design-Implementation-Guide.md [hunk 3]
```diff
@@ -0,0 +1,1979 @@
...
+* 不应该包含任何**业务逻辑**。
+* **切勿**继承或引用**实体**。
+
+**输入DTOs**（应用服务方法的参数）与 **输出DTO** （应用服务方法的返回对象）具有不同的作用，因此，它们应该区别对待。
```
Comment (rqx110) [inline docs/zh-Hans/Domain-Driven-Design-Implementation-Guide.md]:
```suggestion
**输入DTOs**（应用服务方法的参数）与 **输出DTOs** （应用服务方法的返回对象）具有不同的作用，因此，它们应该区别对待。 
```

File: docs/zh-Hans/Domain-Driven-Design-Implementation-Guide.md [hunk 4]
```diff
@@ -0,0 +1,1979 @@
...
+
+**仅**在输入DTOs中定义用例**所需要的属性**！否则，会造成调用应用服务的客户端产生困惑。
+
+这个规则好像没什么必要，因为没人会在方法参数（输入DTOs）中添加无用的属性。但是，有时候，特别是在重用DTO时，输入DTOs会包含无用的属性。
+
+##### 不要重用输入DTOs
+
+**为每个用例**（应用服务的方法）单独定义一个**专属的输入DTO**。否则，在一些情况下，会添加一些不被使用的属性，这样就违反上面的规则：不要在输入DTOs中定义不使用的属性。
+
+在两个用例中重用相同的DTO似乎很有吸引力，因为它们的属性是一模一样的。现阶段它们是一样的，但是随着业务变化，可能它们会产生差异，届时您可能还是需要进行拆分。**代码的重复相对用例间的耦合，可能是更好的做法**。
```
Comment (rqx110) [inline docs/zh-Hans/Domain-Driven-Design-Implementation-Guide.md]:
```suggestion
在两个用例中重用相同的DTO似乎很有吸引力，因为它们的属性是一模一样的。现阶段它们是一样的，但是随着业务变化，可能它们会产生差异，届时您可能还是需要进行拆分。**和用例间的耦合相比，代码的复制可能是更好的做法**。
```

File: docs/zh-Hans/Domain-Driven-Design-Implementation-Guide.md [hunk 5]
```diff
@@ -0,0 +1,1979 @@
...
+}
+````
+
+当输入无效时，ABP框架会自动验证输入DTO，抛出`AbpValidationException`异常，并向客户返回`400`的HTTP状态码。
+
+> 一些开发人员认为最好将验证规则和DTO分离。我们认为声明性（数据注解）方式是比较实用的，不会引起任何设计问题。如果您喜欢其它方式，ABP还支持[FluentValidation继承](FluentValidation.md)。有关所有验证的详细文档，请参见[验证文档](Validation.md)。
+
+#### 输出DTO最佳实践
+
+* 保持输出DTO的**个数较少**，尽可能**重用输入DTO**（例外：不要将输入DTO作为输出DTO）。
```
Comment (rqx110) [inline docs/zh-Hans/Domain-Driven-Design-Implementation-Guide.md]:
```suggestion
* 保持**数量较少**的输出DTO，尽可能**重用输出DTO**（例外：不要将输入DTO作为输出DTO）。
```

File: docs/zh-Hans/Domain-Driven-Design-Implementation-Guide.md [hunk 6]
```diff
@@ -0,0 +1,1979 @@
...
+* 输入DTO可以包含比**用例需要的属性更多的属性**。
```
Comment (rqx110) [inline docs/zh-Hans/Domain-Driven-Design-Implementation-Guide.md]:
```suggestion
* 输出DTO可以包含比用例需要的属性**更多**的属性。
```

File: docs/zh-Hans/Domain-Driven-Design-Implementation-Guide.md [hunk 7]
```diff
@@ -0,0 +1,1979 @@
...
+
+输出DTO的建议并不适用于所有情况。出于**性能**原因，我们可以忽略这些建议，尤其是在返回**大量数据**、为UI定制、**并发量较高**时。
+
+在这些情况下，您可以定制仅包含**必要信息的DTO**。上面的建议只适用于额外多些属性并**不会损失太多性能**，并关注代码**可维护**的应用系统。
+
+#### 对象映射到对象
+
+当两个对象具有相同或相似的属性，自动将[对象映射到对象](Object-To-Object-Mapping.md)是一种将值从一个对象复制到另外一个对象非常有用的方法。
+
+DTO和实体通常具有相同或相似的属性，您经常需要从一个实体创建一个DTO对象。于手动映射相比，基于[AutoMapper](http://automapper.org/)的ABP[对象映射系统](Object-To-Object-Mapping.md)，更加方便简单。
```
Comment (rqx110) [inline docs/zh-Hans/Domain-Driven-Design-Implementation-Guide.md]:
```suggestion
DTO和实体通常具有相同或相似的属性，您经常需要从一个实体创建一个DTO对象。相较于手动映射，基于[AutoMapper](http://automapper.org/)的ABP[对象映射系统](Object-To-Object-Mapping.md)，更加方便简单。
```

File: docs/zh-Hans/Domain-Driven-Design-Implementation-Guide.md [hunk 8]
```diff
@@ -0,0 +1,1979 @@
...
+
+* **仅**在**实体=>输出DTO**的时候使用自动映射。**Use** auto object mapping only for **Entity to output DTO** mappings.
```
Comment (rqx110) [inline docs/zh-Hans/Domain-Driven-Design-Implementation-Guide.md]:
```suggestion
* **仅**在**实体=>输出DTO**的时候使用自动映射。
```

File: docs/zh-Hans/Domain-Driven-Design-Implementation-Guide.md [hunk 9]
```diff
@@ -0,0 +1,1979 @@
...
+* 不要在**输入DTO=>Entity**的时候自动自动映射。
```
Comment (rqx110) [inline docs/zh-Hans/Domain-Driven-Design-Implementation-Guide.md]:
```suggestion
* 不要在**输入DTO=>Entity**的时候使用自动映射。
```

File: docs/zh-Hans/Domain-Driven-Design-Implementation-Guide.md [hunk 10]
```diff
@@ -0,0 +1,1979 @@
...
+
+假如在`IssueManager.CreateAsync`中先保存一次数据，那么数据会先执行一次*Insert*操作，后面关联用户的逻辑执行后，又会再执行一次*Update*操作。
+
+如果不在`IssueManager.CreateAsync`中保存数据，那么，新建`Issue`和关联用户，只会执行一次*Insert*操作。
+
+##### 讨论：为什么没有在应用服务中执行标题是否重复的检查?
+
+简单地说“因为它是**核心领域逻辑**，应该在领域层实现”。这又带来一个新问题，“**如何确定**是领域层逻辑，还是应用层逻辑”？（这个我们后面再详细讨论）
+
+对于此示例，可以用一个简单的问题来判断到底是领域逻辑还是应用逻辑：“如果又另外一种创建`Issue`的方式（用例），我们是否还需要执行？如果需要执行，就属于领域层逻辑，不需要执行就是应用层逻辑”。您可能认为为什么还有别的用例来创建`Issue`呢？
```
Comment (rqx110) [inline docs/zh-Hans/Domain-Driven-Design-Implementation-Guide.md]:
```suggestion
对于此示例，可以用一个简单的问题来判断到底是领域逻辑还是应用逻辑：“如果还有另外一种创建`Issue`的方式（用例），我们是否还需要执行？如果需要执行，就属于领域层逻辑，不需要执行就是应用层逻辑”。您可能认为为什么还有别的用例来创建`Issue`呢？
```

File: docs/zh-Hans/Domain-Driven-Design-Implementation-Guide.md [hunk 11]
```diff
@@ -0,0 +1,1979 @@
...
+        // Update entity in the database
+        await _issueRepository.UpdateAsync(issue);
+
+        // Return a DTO represents the new Issue
+        return ObjectMapper.Map<Issue, IssueDto>(issue);
+    }
+}
+````
+
+* `UpdateAsync` 方法参数 `id`被作为独立参数，放置在`UpdateIssueDto`之外，当您将此应用服务[自动导出](API/Auto-API-Controllers.md)为HTTP API时，可以帮忙ABP准确判断路由，这与DDD无关。
```
Comment (rqx110) [inline docs/zh-Hans/Domain-Driven-Design-Implementation-Guide.md]:
```suggestion
* `UpdateAsync` 方法参数 `id`被作为独立参数，放置在`UpdateIssueDto`之外。这是一项设计决策，可在您将服务[自动导出](API/Auto-API-Controllers.md)为HTTP API端点时帮助ABP正确定义HTTP路由。因此这与DDD无关。
```

File: docs/zh-Hans/Domain-Driven-Design-Implementation-Guide.md [hunk 12]
```diff
@@ -0,0 +1,1979 @@
...
+````
+
+该应用服务本身**不执行任何操作**，并将所有**操作转发给** *领域服务*。它甚至将DTO传递给`IssueManager`
+
+* 如果没有**任何业务逻辑**，只有简单的**CRUD**操作，**请勿**创建领域服务。
+* **切勿**将**DTOs**传递给领域服务，或从领域服务返回**DTOs**。
+
+可以在应用服务中直接注入仓储，实现查询、创建、更新及删除操作。除非在这些操作过程中需要执行某些业务逻辑，在这种情况下，请创建领域服务。
+
+> 不要创建这种CRUD领域服务([YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it))，等将来确实需要时，我们再进行重构，重构并不会影响UI或其它客户端。
```
Comment (rqx110) [inline docs/zh-Hans/Domain-Driven-Design-Implementation-Guide.md]:
```suggestion
> 不要创建“将来可能需要”这种CRUD域服务方法([YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it))！ 在需要时重构它并重构现有代码。 由于应用层优雅地抽象了领域层，因此重构过程不会影响UI层和其他客户端。
```

File: docs/zh-Hans/Domain-Driven-Design-Implementation-Guide.md [hunk 13]
```diff
@@ -0,0 +1,1979 @@
...
+* 在`Web`项目中移除对`Application`和`EntityFrameworkCore`的引用，`Web`在启动时，再动态加载程序集``IssueTracking.Application.dll`和`IssueTracking.EntityFrameworkCore.dll`。可以使用ABP框架的[插件模块](PlugIn-Modules.md)来动态加载程序集。
+
+### DDD模式的应用程序执行顺序
+
+下图展示了基于DDD模式下的Web应用程序执行顺序：
+
+![](images/domain-driven-design-web-request-flow.png)
+
+* 通常由UI（用例）发起一个HTTP请求到服务器。
+* 由展现层（或分布式服务层）中的一个*MVC Controller*或*Razor Page Handler*处理请求，在这次阶段可以执行一些AOP逻辑（授权，验证，异常处理等），*MVC Controller*或*Razor Page Handler*调用注入的应用服务接口，并返回其调用后的结果（DTOs）。.
```
Comment (rqx110) [inline docs/zh-Hans/Domain-Driven-Design-Implementation-Guide.md]:
```suggestion
* 由展现层（或分布式服务层）中的一个*MVC Controller*或*Razor Page Handler*处理请求，在这个阶段可以执行一些AOP逻辑（授权，验证，异常处理等），*MVC Controller*或*Razor Page Handler*调用注入的应用服务接口，并返回其调用后的结果（DTOs）。.
```

**Comment by hikalkan** [comment]
OMG! That's a great effort! Thanks @akinix and @rqx110 :)

**Review by realLiangshiwei — APPROVED** [review]

**Pull request merged**